clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -384,6 +384,16 @@ boolean joinGroupIfNeeded(final Timer timer) {
                 return false;
             }
 
+            Generation gen;
+
+            // Generation data maybe concurrently cleared by Heartbeat thread.
+            // Can't use synchronized for {@code onJoinComplete}, because it can be long enough
+            // and  shouldn't block hearbeat thread.
+            // See {@link PlaintextConsumerTest#testMaxPollIntervalMsDelayInRevocation
+            synchronized (this) {
```

Comment:
Related to my other comment below: in the first step of sending join-group request, even if the generation was reset in the middle that is fine as well since we only use its member id in `.setMemberId(gen.memberId)` and inside `onJoinPrepare`. So at that time if it is reset, then just sending an UNKNOWN member-id is legal. I think we can just remove the changes here.